src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -407,5 +406,176 @@ public static void SetLastWriteTime(string fullPath, DateTimeOffset time, bool a
+            if (data.dwFileAttributes != -1 &&
+                isDirectory != ((data.dwFileAttributes & Interop.Kernel32.FileAttributes.FILE_ATTRIBUTE_DIRECTORY) != 0))
+            {
+                throw new IOException(SR.IO_InconsistentLinkType);
+            }
+
+            bool result = Interop.Kernel32.CreateSymbolicLink(path, pathToTarget, isDirectory);
+            if (!result)
+            {
+                throw Win32Marshal.GetExceptionForLastWin32Error(path);
+            }
+        }
+
+        internal static unsafe FileSystemInfo? ResolveLinkTarget(string linkPath, bool returnFinalTarget, bool isDirectory)
+        {
+            string? targetPath = returnFinalTarget ?
+                GetFinalLinkTarget(linkPath, isDirectory) :
+                GetImmediateLinkTarget(linkPath, isDirectory, throwOnNotFound: true, normalize: true);
+
+            return targetPath == null ? null :
+                isDirectory ? new DirectoryInfo(targetPath) : new FileInfo(targetPath);
+        }
+
+        internal static string? GetLinkTarget(string linkPath, bool isDirectory)
+            => GetImmediateLinkTarget(linkPath, isDirectory, throwOnNotFound: false, normalize: false);
+
+        /// <summary>
+        /// Gets reparse point information associated to <paramref name="linkPath"/>.
+        /// </summary>
+        /// <returns>The immediate link target, absolute or relative or null if the file is not a supported link.</returns>
+        internal static unsafe string? GetImmediateLinkTarget(string linkPath, bool isDirectory, bool throwOnNotFound, bool normalize)
+        {
+            using SafeFileHandle handle = OpenSafeFileHandle(linkPath,
+                    Interop.Kernel32.FileOperations.FILE_FLAG_BACKUP_SEMANTICS |
+                    Interop.Kernel32.FileOperations.FILE_FLAG_OPEN_REPARSE_POINT);
+
+            if (handle.IsInvalid)
+            {
+                int error = Marshal.GetLastWin32Error();
+                switch (error)
+                {
+                    case Interop.Errors.ERROR_FILE_NOT_FOUND:
+                    case Interop.Errors.ERROR_PATH_NOT_FOUND:
+                        if (throwOnNotFound)
+                        {
+                            throw Win32Marshal.GetExceptionForWin32Error(
+                                // File not found doesn't make much sense coming from a directory.
+                                isDirectory ? Interop.Errors.ERROR_PATH_NOT_FOUND : error, linkPath);
+                        }
+                        return null;
+                    default:
+                        throw Win32Marshal.GetExceptionForWin32Error(error, linkPath);
+                }
+            }
+
+            byte[]? buffer = null;
+            try
+            {
+                buffer = ArrayPool<byte>.Shared.Rent(Interop.Kernel32.MAXIMUM_REPARSE_DATA_BUFFER_SIZE);
+                bool success = Interop.Kernel32.DeviceIoControl(
+                    handle,
```

Comment:
Is NtQueryDirectoryFile capable of returning the reparse point's target as unresolved? i.e: as a relative or absolute path.
If so can you please exemplify how can we use that instead of  DeviceIoControl?

Reply:
Since we open with FILE_FLAG_OPEN_REPARSE_POINT flag I'd expect we get information for symlink itself.
Then we could use [NtQueryInformationFile](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-ntqueryinformationfile) with FileNameInformation for FileInformationClass parameter.
I don't know whether NtQueryInformationFile is faster then DeviceIoControl

Reply:
Keep in mind that this is used by `string? FileSystemInfo.LinkTarget` which specifically returns the link's unresolved target.

I don't quite follow how `NtQueryInformationFile` could allow us to get the link's _unresolved target_ path, using it with `FileNameInformation` returns to me the path to the link _itself_; nevertheless, I looked at the other FileInformationClass values but none of them seem to be able to contain information related to the target. If you could please elaborate on how to do this, and if said function returns an equivalent to what we get using DeviceioControl, I am more than happy to change. 

FWIW here's what I tried: https://gist.github.com/Jozkee/23047316c7173b6a0e19e9635a44a3f4

Reply:
@Jozkee based on your findings and the recent changes to mapping logic (using `PrintName`) I am OK with the current implementation and I am marking this as resolved